pkg/controllers/disruption/consolidation.go

```diff
@@ -198,6 +203,73 @@ func (c *consolidation) computeConsolidation(ctx context.Context, candidates ...
+
+	if len(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions) == 0 {
+		if len(candidates) == 1 {
+			c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "Can't replace spot node with a cheaper spot node")...)
+		}
+		// no instance types remain after filtering by price
+		return Command{}, nil
+	}
+
+	// For multi-node consolidation:
+	// We don't have any requirement to check the remaining instance type flexibility, so exit early in this case.
+	if len(candidates) > 1 {
+		return Command{
+			candidates:   candidates,
+			replacements: results.NewNodeClaims,
+		}, nil
+	}
+
+	// For single-node consolidation:
+	// We check whether we have 15 cheaper instances than the current candidate instance. If this is the case, we know the following things:
+	//   1) The current candidate is not in the set of the 15 cheapest instance types and
+	//   2) There were at least 15 options cheaper than the current candidate.
+	if len(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions) < MinInstanceTypesForSpotToSpotConsolidation {
+		c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, fmt.Sprintf("SpotToSpotConsolidation requires %d cheaper instance type options than the current candidate to consolidate, got %d",
+			MinInstanceTypesForSpotToSpotConsolidation, len(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions)))...)
+		return Command{}, nil
+	}
+
+	// Restrict the InstanceTypeOptions for launch to 15 so we don't get into a continual consolidation situation.
+	// For example:
+	// 1) Suppose we have 5 instance types, (A, B, C, D, E) in order of price with the minimum flexibility 3 and they’ll all work for our pod.  We send CreateInstanceFromTypes(A,B,C,D,E) and it gives us a E type based on price and availability of spot.
+	// 2) We check if E is part of (A,B,C) and it isn't, so we will immediately have consolidation send a CreateInstanceFromTypes(A,B,C,D), since they’re cheaper than E.
+	// 3) Assuming CreateInstanceFromTypes(A,B,C,D) returned D, we check if D is part of (A,B,C) and it isn't, so will have another consolidation send a CreateInstanceFromTypes(A,B,C), since they’re cheaper than D resulting in continual consolidation.
+	// If we had restricted instance types to min flexibility at launch at step (1) i.e CreateInstanceFromTypes(A,B,C), we would have received the instance type part of the list preventing immediate consolidation.
+	// Taking this to 15 types, we need to only send the 15 cheapest types in the CreateInstanceFromTypes call so that the resulting instance is always in that set of 15 and we won’t immediately consolidate.
+	results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions = lo.Slice(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions, 0, MinInstanceTypesForSpotToSpotConsolidation)
```

Comment:
I realize this is a super late comment, but curious why we don't log a message for what we're truncating to? Does the consolidation candidate list already get logged somewhere else?

Reply:
Yeah, you'd be able to see it in the NodeClaim requirements. You'd also be able to see it in the generic log message that disruption prints out here which includes the instance types: https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/disruption/controller.go#L189